lib/aot/compiler.cpp

```diff
@@ -2905,6 +3171,148 @@ class FunctionCompiler {
+    auto *Load = Builder.CreateLoad(TargetType, Ptr, OptNone);
+    Load->setAlignment(Align(UINT64_C(1) << Alignment));
+    Load->setAtomic(llvm::AtomicOrdering::SequentiallyConsistent);
+
+    if (Signed) {
+      Stack.back() = Builder.CreateSExt(Load, IntType);
+    } else {
+      Stack.back() = Builder.CreateZExt(Load, IntType);
+    }
+  }
+  void compileAtomicStore(unsigned MemoryIndex, unsigned Alignment,
+                          llvm::IntegerType *, llvm::IntegerType *TargetType,
+                          bool Signed = false) {
+
+    auto *V = stackPop();
+
+    if (Signed) {
+      V = Builder.CreateSExtOrTrunc(V, TargetType);
+    } else {
+      V = Builder.CreateZExtOrTrunc(V, TargetType);
+    }
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
+    compileAtomicCheckOffsetAlignment(Offset, TargetType);
+    auto *VPtr = Builder.CreateInBoundsGEP(
+        Context.Int8Ty, Context.getMemory(Builder, ExecCtx, MemoryIndex),
+        Offset);
+    auto *Ptr = Builder.CreateBitCast(VPtr, TargetType->getPointerTo());
+    auto *Store = Builder.CreateStore(V, Ptr, OptNone);
+    Store->setAlignment(Align(UINT64_C(1) << Alignment));
+    Store->setAtomic(llvm::AtomicOrdering::SequentiallyConsistent);
+  }
+
+  void compileAtomicRMWOp(unsigned MemoryIndex,
+                          [[maybe_unused]] unsigned Alignment,
+                          llvm::AtomicRMWInst::BinOp BinOp,
+                          llvm::IntegerType *IntType,
+                          llvm::IntegerType *TargetType, bool Signed = false) {
+    auto *Value = Builder.CreateSExtOrTrunc(stackPop(), TargetType);
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
+    compileAtomicCheckOffsetAlignment(Offset, TargetType);
+    auto *VPtr = Builder.CreateInBoundsGEP(
+        Context.Int8Ty, Context.getMemory(Builder, ExecCtx, MemoryIndex),
+        Offset);
+    auto *Ptr = Builder.CreateBitCast(VPtr, TargetType->getPointerTo());
+
+    Stack.back() =
+        Builder.CreateAtomicRMW(BinOp, Ptr, Value,
+#if LLVM_VERSION_MAJOR >= 13
+                                Align(UINT64_C(1) << Alignment),
+#endif
+                                llvm::AtomicOrdering::SequentiallyConsistent);
+
+    if (Signed) {
+      Stack.back() = Builder.CreateSExt(Stack.back(), IntType);
+    } else {
+      Stack.back() = Builder.CreateZExt(Stack.back(), IntType);
+    }
+  }
+  void compileAtomicCompareExchange(unsigned MemoryIndex,
+                                    [[maybe_unused]] unsigned Alignment,
+                                    llvm::IntegerType *IntType,
+                                    llvm::IntegerType *TargetType,
+                                    bool Signed = false) {
+
+    auto *Replacement = Builder.CreateSExtOrTrunc(stackPop(), TargetType);
+    auto *Expected = Builder.CreateSExtOrTrunc(stackPop(), TargetType);
+    auto *Offset = Builder.CreateZExt(Stack.back(), Context.Int64Ty);
```

Comment:
Final offset should add offset value in `instruction`.
offset = 'stack pop value' + 'instruction offset value'